euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,531 @@
+[**Hartree atomic units**](https://en.wikipedia.org/wiki/Hartree_atomic_units).
+This means the units are:
+
+- length: bohr radius
+- mass: electron mass
+- energy/frequency: hartree
+- charge: electron charge
+
+Using `Pint >= 0.10.1` units can be aliased and added to the unit registry which
+allows a record of the default user-facing and internal units to be kept. For
+example `INTERNAL_CELL_VECTORS_UNIT` can be defined:
+```
+ureg.define('@alias bohr = INTERNAL_CELL_VECTORS_UNIT')
+```
+Which is what is used to create the `ureg` in the above Structure example.
+
+Inside Euphonic's data objects, all pieces of data with units are stored as a
+bare magnitude, implicitly in atomic units. For example, in the Structure class
+above `_cell_vectors` is a  'private' attribute which is stored as a plain Numpy
+array in units of bohr radius. The user-facing attribute is `cell_vectors` which
+is a property that converts the `_cell_vectors` to whatever unit is stored in
+the `cell_vectors_unit` attribute (e.g. Angstrom) and outputs a `Quantity` to
+the user which makes the units explicit. The output unit can be changed by
+changing the `cell_vectors_unit` attribute, this could be wrapped as a property
+to check for valid units before being set.
+
+When instantiating an object directly with the constructor, the user is expected
+to provide any data containing units as a `Quantity` specifying the units. The
+default units for that object will then be the same as whatever was input.
+
+When outputting to/taking input from another data format e.g. dictionary or
+file, `Quantity` objects are not used to increase portability. Instead, any data
+with units will be output as 2 fields: 1 containing the data, and 1 containing a
+string specifying the units. For example, in the Structure class above, when
+converting `cell_vectors` to a dictionary, the dictionary will contain:
+```
+'cell_vectors': array([[ 2.426176, -4.20226   0.      ],
+                       [ 2.426176,  4.20226   0.      ],
+                       [ 0.,        0.,       5.350304]])
+'cell_vectors_unit': 'angstrom'
+```
+
+## Plotting
+Plotting functions will be contained in their own separate module, so that
+libraries such as `Matplotlib` will only be imported if a user wants to plot,
+and not required for general data analysis. In addition, a user may want to plot
+several Spectrum objects on the same axes, which requires plotting be done from
+a separate module rather than called as method on the Spectrum object. 
+
+Plotting in Euphonic will be fairly simple, as it is not Euphonic's main goal,
+and will be mainly useful for debugging and quick sanity checking of answers. 
+
+## Serialisation
+For both easy testing and export to other programs, Euphonic will need to be
+able to serialise most of its data objects. They should each have their own
+methods, although it is not decided what formats should be supported and more
+may be added later. Initial support could be for one of: 
+
```

Comment:
Are we still considering HDF5 for large objects?

Reply:
Yes, I think that's our most likely data format for large objects. I just listed some human-readable file types as that's probably what will get implemented first because I'll be using it for testing. I'll add a mention of HDF5.